Interface definition for application defined thread implementation.

temeo opened the pull request:
The interface will allow application to inject service interface into
provider. The service interface will define callbacks to manage
threads, mutexes and conditions variables.

ayurchen commented on wsrep_thread_service.h:
```diff
@@ -0,0 +1,349 @@
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 2 of the License, or
+ * (at your option) any later version.
+ *
```
If we want to have a separate wsrep_loader.h file, it makes sense to put the same wording in there. And probably in wsrep_api.h as well. I mean, to have a common head in all three files.

temeo replied:
Maybe better to update the license header for all files in separate PR.

ayurchen commented on wsrep_thread_service.h:
```diff
@@ -0,0 +1,349 @@
+    typedef struct wsrep_mutex_st wsrep_mutex_t;
+    typedef struct wsrep_cond_key_st wsrep_cond_key_t;
+    /** Condition variable tags */
+    typedef struct wsrep_cond_st wsrep_cond_t;
```
I think here line 63 and line 64 must be swapped. And same goes for 66 and 67

temeo commented:
> What is the motivation to have a separate `wsrep_loader.h`?

This change was not necessary at this point. Reverted the changes to `wsrep_api.h` and removed `wsrep_loader.h`.
